Src/userkybd.cpp

```diff
@@ -44,37 +45,244 @@ static void DrawBorder(HDC hDC, RECT rect, BOOL Depressed);
+	{ IDK_COMMA, ",", 6, 6 },
+	{ IDK_CARET, "^", 1, 8 },
+	{ IDK_DOT, ".", 6, 7 },
+	{ IDK_FWDSLASH, "/", 6, 8 },
+	{ IDK_STAR, "*", 4, 8 },
+	{ IDK_OPEN_SQUARE, "[", 3, 8 },
+	{ IDK_BACKSLASH, "\\", 7, 8 },
+	{ IDK_CLOSE_SQUARE, "]", 5, 8 },
+	{ IDK_AT, "@", 4, 7 },
+	{ IDK_UNDERSCORE, "_", 2, 8 },
+};
+
+const std::pair<const char*, const char*> BBCKeyAliases[] = {
+	{ "DEL", "DELETE" },
+	{ "CAPS", "CAPS-LOCK" },
+	{ "CAPSLOCK", "CAPS-LOCK" },
+	{ "CONTROL", "CTRL" },
+	{ "ESC", "ESCAPE" },
+	{ "SHIFTLOCK", "SHIFT-LOCK" }
 };
 
+// Number of inputs assigned to each key (for joystick dialog highlight)
+static int assignedKeysCount[_countof(BBCKeys)] = {};
+
+/****************************************************************************/
+
+const BBCKey& GetBBCKeyByResId(int ctrlId)
+{
+    using resIdToKeyMapType = std::map<int, const BBCKey*>;
+
+    // Construct map on first use by lambda
+    static resIdToKeyMapType resIdToKeyMap = []()
+    {
+        resIdToKeyMapType keyMap{};
+
+        for (auto& theKey : BBCKeys)
+            keyMap[theKey.ctrlId] = &theKey;
+        return keyMap;
+    } ();
+
+    auto iter = resIdToKeyMap.find(ctrlId);
+    if (iter == resIdToKeyMap.end())
+        return BBCKeys[0];
+
+    return *iter->second;
+}
+
+/****************************************************************************/
+
+const BBCKey& GetBBCKeyByName(const std::string& name)
+{
+    using nameToKeyMapType = std::map<std::string, const BBCKey*>;
+
+    // Construct map on first use by lambda
+    static nameToKeyMapType nameToKeyMap = []()
+    {
+        nameToKeyMapType keyMap{};
+
+        for (auto& theKey : BBCKeys)
+            keyMap[theKey.name] = &theKey;
+
+        for (auto& alias : BBCKeyAliases)
+            keyMap[alias.first] = keyMap[alias.second];
+
+        return keyMap;
+    } ();
+
+    auto iter = nameToKeyMap.find(name);
+    if (iter == nameToKeyMap.end())
+        return BBCKeys[0];
+
+    return *iter->second;
+}
+
 /****************************************************************************/
 
-bool UserKeyboardDialog(HWND hwndParent)
+const BBCKey& GetBBCKeyByRowAndCol(int row, int col)
+{
+    using posPair = std::pair<int, int>;
+    using posToKeyMapType = std::map<posPair, const BBCKey*>;
+
+    // Construct map on first use by lambda
+    static posToKeyMapType posToKeyMap = []()
+    {
+        posToKeyMapType keyMap{};
+
+        for (auto& theKey : BBCKeys)
+            keyMap[posPair{ theKey.row, theKey.column }] = &theKey;
+        return keyMap;
+    } ();
+
+    auto iter = posToKeyMap.find(posPair{ row, col });
+    if (iter == posToKeyMap.end())
+        return BBCKeys[0];
+
+    return *iter->second;
+}
+
+/****************************************************************************/
+// Get index for assignedKeysCount. The 'key' parameter must be reference to
+// item in BBCKeys table, not a copy of it.
+int GetBBCKeyIndex(const BBCKey& key)
+{
+    int index = &key - BBCKeys;
```

Comment:
I think this may be clearer if the function takes a `BBCKey` pointer instead of a reference, and the various map functions (`GetBBCKeyByResId`) could return pointers.